src/js_sync.hpp

```diff
@@ -679,6 +687,59 @@ void SessionClass<T>::override_server(ContextType ctx, ObjectType this_object, A
     }
 }
 
+template<typename T>
+void SessionClass<T>::wait_for_completion(Direction direction, ContextType ctx, ObjectType this_object, Arguments &args) {
+    args.validate_count(1);
+    if (auto session = get_internal<T, SessionClass<T>>(this_object)->lock()) {
+        auto callback_function = Value::validated_to_function(ctx, args[0]);
+        Protected<FunctionType> protected_callback(ctx, callback_function);
+        Protected<ObjectType> protected_this(ctx, this_object);
+        Protected<typename T::GlobalContext> protected_ctx(Context<T>::get_global_context(ctx));
+
+        std::function<DownloadUploadCompletionHandler> completion_func;
+        EventLoopDispatcher<DownloadUploadCompletionHandler> completion_handler([=](std::error_code error) {
+            HANDLESCOPE
+            ValueType callback_arguments[1];
+            if (error) {
+                ObjectType error_object = Object::create_empty(protected_ctx);
+                Object::set_property(protected_ctx, error_object, "message", Value::from_string(protected_ctx, error.message()));
+                Object::set_property(protected_ctx, error_object, "errorCode", Value::from_number(protected_ctx, error.value()));
+                callback_arguments[0] = error_object;
+            } else {
+                callback_arguments[0] = Value::from_undefined(ctx);
+            }
+            Function<T>::callback(protected_ctx, protected_callback, typename T::Object(), 1, callback_arguments);
+        });
+
+        completion_func = std::move(completion_handler);
+        bool callback_registered;
+        switch(direction) {
+            case Upload:
+                callback_registered = session->wait_for_upload_completion(std::move(completion_func));
+                break;
+            case Download:
+                callback_registered = session->wait_for_download_completion(std::move(completion_func));
+                break;
+        }
+        if (!callback_registered) {
```

Comment:
Wouldn't it be better to handle that in a `default` case rather than introducing the extra bool variable?

Reply:
The problem is that that registering the listener can apparently fail. At least that is how the sync API is documented, so it isn't an extra case as such.